pallets/asset-registry/src/lib.rs

```diff
@@ -265,181 +346,148 @@ pub mod pallet {
+
+				if decimals.is_some() {
+					if details.decimals.is_none() {
+						details.decimals = decimals;
+					} else {
+						//Only highest origin can change decimal if it was set previously.
+						ensure!(is_registry_origin, Error::<T>::Forbidden);
+						details.decimals = decimals;
+					};
 				}
 
-				detail.name = bounded_name.clone();
-				detail.asset_type = asset_type;
-				detail.existential_deposit = existential_deposit.unwrap_or(detail.existential_deposit);
-				detail.xcm_rate_limit = xcm_rate_limit;
+				if let Some(loc) = location {
+					//Only highest origin can update location.
+					ensure!(is_registry_origin, Error::<T>::Forbidden);
+
+					if let Some(old_location) = AssetLocations::<T>::take(asset_id) {
+						LocationAssets::<T>::remove(&old_location);
+					}
+					Self::do_set_location(asset_id, loc)?;
+				}
 
 				Self::deposit_event(Event::Updated {
 					asset_id,
-					asset_name: bounded_name,
-					asset_type,
-					existential_deposit: detail.existential_deposit,
-					xcm_rate_limit: detail.xcm_rate_limit,
+					asset_name: details.name.clone(),
+					asset_type: details.asset_type,
+					existential_deposit: details.existential_deposit,
+					xcm_rate_limit: details.xcm_rate_limit,
+					symbol: details.symbol.clone(),
+					decimals: details.decimals,
+					is_sufficient: details.is_sufficient,
 				});
 
 				Ok(())
 			})
 		}
 
-		/// Set metadata for an asset.
-		///
-		/// - `asset_id`: Asset identifier.
-		/// - `symbol`: The exchange symbol for this asset. Limited in length by `StringLimit`.
-		/// - `decimals`: The number of decimals this asset uses to represent one unit.
-		///
-		/// Emits `MetadataSet` event when successful.
-		#[pallet::call_index(2)]
-		#[pallet::weight(<T as Config>::WeightInfo::set_metadata())]
-		pub fn set_metadata(
-			origin: OriginFor<T>,
-			asset_id: T::AssetId,
-			symbol: Vec<u8>,
-			decimals: u8,
-		) -> DispatchResult {
-			T::RegistryOrigin::ensure_origin(origin)?;
-
-			ensure!(Self::assets(asset_id).is_some(), Error::<T>::AssetNotFound);
-
-			let b_symbol = Self::to_bounded_name(symbol)?;
-
-			let metadata = AssetMetadata::<BoundedVec<u8, T::StringLimit>> {
-				symbol: b_symbol.clone(),
-				decimals,
-			};
-
-			AssetMetadataMap::<T>::insert(asset_id, metadata);
-
-			Self::deposit_event(Event::MetadataSet {
-				asset_id,
-				symbol: b_symbol,
-				decimals,
-			});
-
-			Ok(())
-		}
-
-		/// Set asset native location.
-		///
-		/// Adds mapping between native location and local asset id and vice versa.
-		///
-		/// Mainly used in XCM.
-		///
-		/// Emits `LocationSet` event when successful.
-		#[pallet::call_index(3)]
-		#[pallet::weight(<T as Config>::WeightInfo::set_location())]
-		pub fn set_location(
-			origin: OriginFor<T>,
-			asset_id: T::AssetId,
-			location: T::AssetNativeLocation,
-		) -> DispatchResult {
-			T::RegistryOrigin::ensure_origin(origin)?;
+		//NOTE: call indices 2 and 3 were used by removed extrinsics.
```

Comment:
i wonder why we cannot reuse those ?

Reply:
I'm not sure, @apopiak pointed out it's better to retire old and introduce new indices

Reply:
It's up to you, but you need to bump the transaction version if you change the parameters of an extrinsic.